lib/charms/prometheus_k8s/v0/prometheus_scrape.py

```diff
@@ -1123,6 +1123,9 @@ def jobs(self) -> list:
 
         scrape_jobs = _dedupe_job_names(scrape_jobs)
 
+        if not self._tool.validate_scrape_jobs(scrape_jobs):
+            return []
```

Comment:
Not sure we want it to silently fail. I would add a `WARNING`.
This is another "juju model" vs "best effort" question.
I think we should emit a custom event the charm observes and blocks.

(Having status per relation would be handy here.)

Reply:
We log an error message (from within `validate_scrape_jobs`).

Reply:
Not sure why a custom event would be useful here. What would we do with it?

Reply:
I don't even know if this is "juju model" vs "best effort". We can do both.

If this gets bumped up inside the loop, we could do almost exactly the kind of validation we [already do with alerts](https://github.com/canonical/prometheus-k8s-operator/blob/main/lib/charms/prometheus_k8s/v0/prometheus_scrape.py#L1199) (and Grafana dashboards, and Loki alerts). 

```python
if static_scrape_jobs:
    if not self._tool.validate_scrape_jobs(static_scrape_jobs):
      # set a message back in relation data so the Provider can see it, and it's isolated per relation
```

This keeps the information in the model (at least the validity/invalidity), the Provider can catch it in [`_on_relation_changed`](https://github.com/canonical/prometheus-k8s-operator/blob/main/lib/charms/prometheus_k8s/v0/prometheus_scrape.py#L1598) (arguably, this code won't even have to change), and block from the Provider side only.

Then we exclude it from the total list (don't extend scrape jobs with bad ones)

Reply:
I am aware that we do this for alert rules and I'm not sure it was actually the best design. The reason I am skeptical of this approach is at the moment, as far as I can tell, *none* of our charms observe `alert_rule_status_changed`. I am sort of struggling to see what the real life use case for this would be.

Reply:
None of _our_ charms do (`cos-config` really should), but the real life use case is reasonably simple. In a large deployment, or one with gitops, or `cos-proxy`, or other, specifying with some level of granularity to a particular provider that "you gave me bad alert rules/dashboards/scrape targets" is far more specific than blocking Prom/Loki/Grafana and hoping admins can sift through a giant bucket of stuff to figure out what's wrong and why, doesn't block the consumer, doesn't leave it with empty rules, etc.

Even if they don't actively listen to the event, `juju show-unit|application` will tell admins why a rule/dashboard/target from `foo/0` isn't visible.

Conversely, in the current patch, pointing `cos-config` to a git repo with bad scrape targets would cause Prometheus to drop _all_ scrape jobs. Just because none of our charms currently listen to this does not mean that the design a problem.

If you are bulk loading data into a database, going through a multi-select in an application/website, or anything else, being as specific as possible about which data is invalid and why is good UX.